[webkitapipy] Support conditional allowlist entries based on OS / SDK version

emw-apple opened the pull request:
#### 6f94b07d3b3b62262b66e3198cedd34485b5d627
<pre>
[webkitapipy] Support conditional allowlist entries based on OS / SDK version
<a href="https://bugs.webkit.org/show_bug.cgi?id=302293">https://bugs.webkit.org/show_bug.cgi?id=302293</a>
<a href="https://rdar.apple.com/164435576">rdar://164435576</a>

Reviewed by Sam Sneddon.

There are some cases where a more general OS or SDK version check is
superior to conditionalizing an allowlist entry based on a
wtf/Platform.h flag:

- Staged API adoption is based on forward-declaring API that has yet to
  appear in the builds of the SDK. Until it does, it will look like SPI
  to the tool. But its usability (and whether or not it is API) is not based
  on a WebKit feature flag, it&apos;s based on the SDK vesrion.

- API from frameworks that WebKit builds with internally (such as
  AuthenticationServices.framework) may have already shipped in new
  SDKs, but when building against older SDKs appears as SPI. Again, the
  SDK is what determines whether the usage looks like SPI.

To support these use cases, add optional &quot;request-os&quot; and &quot;requires-sdk&quot; version
requirement keys to the allowlist format. Each is a list of strings with
the form &quot;&lt;platform&gt; &lt;operator&gt; &lt;version&gt;&quot;. For example:

    [[staging]]
    classes = [MyNewClass]
    requires = [ENABLE_MY_NEW_FEATURE]
    requires-sdk = [&quot;iOS &lt; 26.2&quot;]

Special wildcard syntactic sugar is available for the purpose of
avoiding writing a misleading marketing version. A requirement like:

    iOS &lt; 26.0

may be written as:

    iOS &lt;= 25.*

At parse time the wildcard expression is transformed into its logical
equivalent.

In the SDKDB cache, this is implemented by extending the existing
&quot;conditional chain&quot; mechanism used for Platform flags. Conditionals now
have a value (WTF-style defines are all value &quot;1&quot;), and each link in the
chain has an associated operator.

* Configurations/CommonBase.xcconfig:
* Source/JavaScriptCore/Configurations/AllowedSPI-legacy.toml:
* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
(AllowedSPI):
(AllowedSPI.RequiredVersion):
(_transform_wildcard_version):
(AllowList.from_dict):

* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow_unittest.py:
(TestAllowList.test_repeated_requirements):
(TestAllowList.test_no_string):
(TestAllowList):
(TestAllowList.test_invalid_version_requirements):
(TestAllowList.test_required_fields):

(TestAllowList.test_wildcard_version):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py: Read the
  platform and versions from each binary&apos;s Mach-O header.
(APIReport):
(APIReport.from_binary):
(APIReport._from_dyld_info):
(APIReport._populate_from_dyld_info):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
(apply_operator_sql):
(semver_to_int): Used to store version numbers as 6-digit integers, for
  easy comparison, similar to what OS headers do.
(SDKDB._initialize_db):
(SDKDB._initialize_temporary_schema):
(SDKDB._add_allowlist):
(SDKDB.add_conditions): New entrypoint for add_defines which takes
  key-value pairs instead of just keys.
(SDKDB._add_imports):
(SDKDB.audit):
(SDKDB.add_defines): Deleted.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_unittest.py:
(TestSDKDB.test_entries_removed_when_binary_updated):
(TestSDKDB.test_audit_allowed_conditional):
(TestSDKDB.test_audit_missing_name_conditional):
(TestSDKDB.test_audit_missing_name_negated_conditional):
(TestSDKDB.test_audit_allowed_negated_conditional):
(TestSDKDB.test_audit_allowed_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions_negation):
(TestSDKDB):
(TestSDKDB.test_audit_minos_conditions):
(TestSDKDB.test_audit_api_in_loaded_and_unloaded_library):
(TestSDKDB.test_audit_allow_different_fully_qualified_methods_same_name):
(TestSDKDB.test_audit_unnecessary_allow_unqualified_methods_same_name):

Canonical link: <a href="https://commits.webkit.org/308228@main">https://commits.webkit.org/308228@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/55f8a1ff562de9e3527ace3cc15d08730068c903

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/146842 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/19523 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/13061 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/155526 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/100231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/148717 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/19982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/19424 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/113151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/100231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/149805 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/19982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/131933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/93904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/146167 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/19982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/12403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk3-libwebrtc~~](https://ews-build.webkit.org/#/builders/173/builds/2967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/138811 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/19982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/167/builds/9802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/157857 "Built successfully") | | 
| [  ~~🛠 🧪 jsc-debug-arm64~~](https://ews-build.webkit.org/#/builders/171/builds/7631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-safer-cpp~~](https://ews-build.webkit.org/#/builders/174/builds/987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/11240 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/121170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/19325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/16232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/121372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/31092 "Built successfully and passed tests") | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/19334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/131556 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/75178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/170/builds/8438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/178131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/18940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/82695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/178131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/18670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/18821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/18729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
```diff
@@ -41,6 +42,7 @@ def __str__(self):
 else:
     from enum import StrEnum
 
+VERSION_REQ = re.compile(r'(?P<platform>[a-zA-Z]+) ?(?P<op>==|!=|>|>=|<=|<) ?(?P<version>\d+(\.\d+)*)')
```
Do we actually want `\d` here? Like, it would be fine if we had `re.ASCII` as a flag, but otherwise it allows any Unicode decimal digit.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -62,14 +65,17 @@ def from_binary(cls, binary_path: Path, *, arch: str,
 
-        report = cls(file=binary_path, arch=arch)
+        report = cls(file=binary_path, arch=arch,
+                     platform='', min_os='', sdk='')
```
I don't love this, because it ends up with me feeling like "why don't we just give them empty strings as a default value".

I might be more tempted to pass `binary_path` and `arch` into `_populate_from_dyld_info` and make it a class method which does the actual construction?

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
```diff
@@ -137,6 +177,7 @@ def __init__(self, db_file: Path):
                 self.con.close()
                 db_file.unlink()
         self._initialize_temporary_schema()
+        self.con.create_function("ApplyOperator", 3, apply_operator, deterministic=True)
```
I don't love this, especially when it's so simple, and especially when it means roundtrips between SQLite and Python.

It makes me sad that SQLite doesn't support `CREATE FUNCTION`, but we could always define a helper like:

```python
def apply_operator_sql(op_col: str, value_col: str, operand_col: str) -> str:
    return (
        f"(CASE ({op_col})"
        f" WHEN '='  THEN ({value_col}) = ({operand_col})"
        f" WHEN '!=' THEN ({value_col}) != ({operand_col})"
        f" WHEN '<'  THEN ({value_col}) < ({operand_col})"
        f" WHEN '<=' THEN ({value_col}) <= ({operand_col})"
        f" WHEN '>'  THEN ({value_col}) > ({operand_col})"
        f" WHEN '>=' THEN ({value_col}) >= ({operand_col})"
        f" ELSE RAISE(ABORT, 'unknown operator')"
        f" END)"
    )
```

emw-apple replied:
We have to use IS / IS NOT instead of = / != to get the correct null semantics, but otherwise yeah this works fine. Thank you!

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
```diff
@@ -159,7 +200,7 @@ def _initialize_db(self):
                     '                       ON DELETE CASCADE)')
         cur.execute('CREATE INDEX allow_names ON allow (name, kind)')
-        cur.execute('CREATE TABLE condition_chain(name, invert, nextid, '
+        cur.execute('CREATE TABLE condition_chain(name, op Operator, operand, nextid, '
```
1. `Operator` is redundant, given it's just treated as an unknown type, no?

2. Do we want to add a `CHECK` constraint like `CHECK (op IN ('=', '!=', '<', '<=', '>', '>='))`?

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
```diff
@@ -64,6 +65,44 @@ def from_sql(cls, value: bytes) -> DeclarationKind:
+        raise ValueError(op)
+
+
+def semver_to_int(semver: str) -> int:
+    version = tuple(map(int, semver.split('.')))
+    if len(version) > 3 or any(x > 99 for x in version):
+        raise ValueError(f'Semantic version "{semver}" must be no more than '
+                         '3 components, with each component below 99')
+    elif len(version) == 3:
+        return version[0] * 10000 + version[1] * 100 + version[2]
+    elif len(version) == 2:
+        return version[0] * 10000 + version[1] * 100
+    else:
+        return version[0] * 10000
```
I feel like it's more normal to allow `0 >= x > 256` for each component (note the lower bound!), and then just treat each as a byte and do everything with bitwise operations?

emw-apple replied:
As an implementation detail it would be fine to do it this way, but I do like keeping it in base 10 for debuggability and to match the same semantics that the OS availability headers use.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
```diff
@@ -86,6 +125,7 @@ class UnnecessaryAllowedName(NamedTuple):
 
 
 Diagnostic = Union[MissingName, UnusedAllowedName, UnnecessaryAllowedName]
+ConditionVariable = Union[bool, str, int, float]
```
It would be nice to not do this, but I guess what we really want is a TypedDict with `extra_items` but that's Python 3.15 or a dependency on `typing_extensions`.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
```diff
@@ -370,15 +424,23 @@ def _add_allowlist(self, config: AllowList, allowlist: Path):
                                         allow_unused=entry.allow_unused)
 
-    def add_defines(self, defines: list[str]):
+    def add_conditions(self, conditions: dict[str, ConditionVariable]):
```
(It would be better practice in general to use `Mapping` rather than `dict` here — because we don't actually care what type of mapping it is, and we don't need it to be mutable.)

gsnedders reviewed (changes requested):
I think my comments on `VERSION_REQ` and the `APIReport` dataclass are the ones that most need addresses here; the rest is a lot less important.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
It’s non-obvious to me why we’re using Decimal here, when we otherwise just use a tuple of ints?

Like it works, it’s just a bit surprising.

emw-apple replied:
I think it's basically a choice between using Decimal and implementing the subset of its functionality we need (which is: a tuple of ints, a floating point, and an adder). I implemented it by hand at first and then found Decimal to be much less code.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
Hypothetically, if we had:

```
-platform:
platform     minOS      sdk
-platform:
platform     minOS      sdk
     iOS     26.0      26.0
```

Then we'd parse this wrong… but I guess we're trusting `dyld` doesn't give us weird output.

emw-apple replied:
There is probably value in adding some tests that specifically defend against `dyld_info` changing its output, but that's out of scope for today.

gsnedders reviewed (approved):
Please fix the style checker failures before landing:

```
Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:189: error: [pep8/E265] block comment should start with '# '
Tools/Scripts/libraries/webkitapipy/webkitapipy/allow_unittest.py:167: error: [pep8/E265] block comment should start with '# '
```

I'd also like you to give at least some rational, at least on the PR, for the below:

emw-apple committed:
[webkitapipy] Support conditional allowlist entries based on OS / SDK version
https://bugs.webkit.org/show_bug.cgi?id=302293
rdar://164435576

Reviewed by Sam Sneddon.

There are some cases where a more general OS or SDK version check is
superior to conditionalizing an allowlist entry based on a
wtf/Platform.h flag:

- Staged API adoption is based on forward-declaring API that has yet to
  appear in the builds of the SDK. Until it does, it will look like SPI
  to the tool. But its usability (and whether or not it is API) is not based
  on a WebKit feature flag, it's based on the SDK vesrion.

- API from frameworks that WebKit builds with internally (such as
  AuthenticationServices.framework) may have already shipped in new
  SDKs, but when building against older SDKs appears as SPI. Again, the
  SDK is what determines whether the usage looks like SPI.

To support these use cases, add optional "request-os" and "requires-sdk" version
requirement keys to the allowlist format. Each is a list of strings with
the form "<platform> <operator> <version>". For example:

    [[staging]]
    classes = [MyNewClass]
    requires = [ENABLE_MY_NEW_FEATURE]
    requires-sdk = ["iOS < 26.2"]

Special wildcard syntactic sugar is available for the purpose of
avoiding writing a misleading marketing version. A requirement like:

    iOS < 26.0

may be written as:

    iOS <= 25.*

At parse time the wildcard expression is transformed into its logical
equivalent.

In the SDKDB cache, this is implemented by extending the existing
"conditional chain" mechanism used for Platform flags. Conditionals now
have a value (WTF-style defines are all value "1"), and each link in the
chain has an associated operator.

* Configurations/CommonBase.xcconfig:
* Source/JavaScriptCore/Configurations/AllowedSPI-legacy.toml:
* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
(AllowedSPI):
(AllowedSPI.RequiredVersion):
(_transform_wildcard_version):
(AllowList.from_dict):

* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow_unittest.py:
(TestAllowList.test_repeated_requirements):
(TestAllowList.test_no_string):
(TestAllowList):
(TestAllowList.test_invalid_version_requirements):
(TestAllowList.test_required_fields):

(TestAllowList.test_wildcard_version):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py: Read the
  platform and versions from each binary's Mach-O header.
(APIReport):
(APIReport.from_binary):
(APIReport._from_dyld_info):
(APIReport._populate_from_dyld_info):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
(apply_operator_sql):
(semver_to_int): Used to store version numbers as 6-digit integers, for
  easy comparison, similar to what OS headers do.
(SDKDB._initialize_db):
(SDKDB._initialize_temporary_schema):
(SDKDB._add_allowlist):
(SDKDB.add_conditions): New entrypoint for add_defines which takes
  key-value pairs instead of just keys.
(SDKDB._add_imports):
(SDKDB.audit):
(SDKDB.add_defines): Deleted.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_unittest.py:
(TestSDKDB.test_entries_removed_when_binary_updated):
(TestSDKDB.test_audit_allowed_conditional):
(TestSDKDB.test_audit_missing_name_conditional):
(TestSDKDB.test_audit_missing_name_negated_conditional):
(TestSDKDB.test_audit_allowed_negated_conditional):
(TestSDKDB.test_audit_allowed_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions_negation):
(TestSDKDB):
(TestSDKDB.test_audit_minos_conditions):
(TestSDKDB.test_audit_api_in_loaded_and_unloaded_library):
(TestSDKDB.test_audit_allow_different_fully_qualified_methods_same_name):
(TestSDKDB.test_audit_unnecessary_allow_unqualified_methods_same_name):

Canonical link: https://commits.webkit.org/308228@main
